onadata/apps/api/viewsets/open_data_viewset.py

```diff
@@ -16,22 +17,92 @@
 from onadata.apps.logger.models import Instance, XForm
 from onadata.apps.logger.models.open_data import OpenData
 from onadata.libs.data import parse_int
+from onadata.libs.utils.common_tags import METADATA_FIELDS
 from onadata.libs.mixins.cache_control_mixin import CacheControlMixin
 from onadata.libs.mixins.etags_mixin import ETagsMixin
 from onadata.libs.pagination import StandardPageNumberPagination
-from onadata.libs.serializers.data_serializer import DataInstanceSerializer
+from onadata.libs.serializers.data_serializer import TableauDataSerializer
 from onadata.libs.serializers.open_data_serializer import OpenDataSerializer
 from onadata.libs.utils.common_tools import json_stream
-from onadata.libs.utils.csv_builder import CSVDataFrameBuilder
+from onadata.libs.utils.common_tags import (ATTACHMENTS, NOTES, GEOLOCATION)
 
 BaseViewset = get_baseviewset_class()
 IGNORED_FIELD_TYPES = ['select one', 'select multiple']
 
+# index tags
+DEFAULT_OPEN_TAG = '['
+DEFAULT_CLOSE_TAG = ']'
+DEFAULT_INDEX_TAGS = (DEFAULT_OPEN_TAG, DEFAULT_CLOSE_TAG)
+
 
 def replace_special_characters_with_underscores(data):
     return [re.sub(r"\W", r"_", a) for a in data]
 
 
+def clean_data_fields(data, xform):
+    """
+    Streamlines the row header fields
+    with the column header fields for the same form.
+    """
+    result = []
+    if data:
+        headers = xform.get_headers()
+        tableau_headers = remove_metadata_fields(headers)
+        # Yield the rows
+        rows = (row for row in data)
+        for row in rows:
+            # Include fields from the column headers
+            diff = set(tableau_headers).difference(set(row))
+            left_out_fields = {
+                field: None for field in diff}
+            row.update(left_out_fields)
+            result.append(row)
+    yield result
```

Comment:
This is a bit nit-picky but I don't see why we should have `result` here.... We don't really need to return a list with one row incremented I believe ?

```suggestion
    if data:
        headers = xform.get_headers()
        tableau_headers = remove_metadata_fields(headers)
        # Yield the rows
        rows = (row for row in data)
        for row in rows:
            # Include fields from the column headers
            diff = set(tableau_headers).difference(set(row))
            left_out_fields = {
                field: None for field in diff}
            row.update(left_out_fields)
            yield row
```